File: content/snippets/typescript/language_basics/interfaces_and_polymorphism.ts
#!/usr/bin/env ts-node

import { setImmediate } from "node:timers";

interface Speaker {
  speak(): string;
}

class Dog implements Speaker {
  speak(): string {
    return "woof";
  }
}

class Cat implements Speaker {
  speak(): string {
    return "meow";
  }
}

function main(): void {
  const speakers: Speaker[] = [new Dog(), new Cat()];
  for (const speaker of speakers) {
    console.log(speaker.speak());
  }
}

main();
setImmediate(() => {});

File: content/snippets/cpp/language_basics/interfaces_and_polymorphism.cpp
#include <iostream>
#include <memory>
#include <string>
#include <vector>

class Speaker {
public:
    virtual ~Speaker() = default;
    virtual std::string speak() const = 0;
};

class Dog : public Speaker {
public:
    std::string speak() const override { return "woof"; }
};

class Cat : public Speaker {
public:
    std::string speak() const override { return "meow"; }
};

int main() {
    std::vector<std::unique_ptr<Speaker>> speakers;
    speakers.push_back(std::make_unique<Dog>());
    speakers.push_back(std::make_unique<Cat>());

    for (const auto& speaker : speakers) {
        std::cout << speaker->speak() << "\n";
    }
    return 0;
}
